### DISCUSSION TITLE
Topic-Word Distributions from approximate_distribution

### DISCUSSION THREAD

**Discussion by pjirach** [MaartenGr/BERTopic#2084] [Q&A]
Hello!

I'd like to know the result when we get the topic-word distributions from topic_model.approximate_distribution(docs, calculate_tokens=True).

Refer from [https://maartengr.github.io/BERTopic/getting_started/distribution/distribution.html](https://maartengr.github.io/BERTopic/getting_started/distribution/distribution.html):
1. How do we calculate topic-word distributions? In simpler terms, do we just calculate the similarity between the topic representation and token representation and apply some function to make it become a probability?

2. I understand that in the default pipeline, the document is tokenized by the tokenizer defined by CountVectorizer before calculating the approximated probability above. I found that, even if we set hyperparameters like stop_words="english", min_df=3, max_df=0.9, we still get some repetitive tokens with different probability. Moreover, we could get two tokens with different capitalization or verb forms (like _s or _ed). How should I fix this?
![image](https://github.com/user-attachments/assets/b718cbc9-61b2-4dd8-a61c-a510596b5280)
I really appreciate your work. Thanks in advance!

**Comment by MaartenGr** [comment]
> How do we calculate topic-word distributions? In simpler terms, do we just calculate the similarity between the topic representation and token representation and apply some function to make it become a probability?

As shown in the visualization, it calculates the similarity between a certain tokenset and the topic representation. For example, this can be the cosine similarity between the c-TF-IDF representations of a tokenset and the topic representation. They are aggregated and normalized:

https://github.com/MaartenGr/BERTopic/blob/bf1fedd5bb14bbf282d5051e849d2b06abaa8b51/bertopic/_bertopic.py#L1382

I would advise checking out the source code as each step is documented:

https://github.com/MaartenGr/BERTopic/blob/bf1fedd5bb14bbf282d5051e849d2b06abaa8b51/bertopic/_bertopic.py#L1182

> I understand that in the default pipeline, the document is tokenized by the tokenizer defined by CountVectorizer before calculating the approximated probability above. I found that, even if we set hyperparameters like stop_words="english", min_df=3, max_df=0.9, we still get some repetitive tokens with different probability. Moreover, we could get two tokens with different capitalization or verb forms (like _s or _ed). How should I fix this?

I believe that might be a result of the `build_tokenizer` that is being used to tokenize the documents:

https://github.com/MaartenGr/BERTopic/blob/bf1fedd5bb14bbf282d5051e849d2b06abaa8b51/bertopic/_bertopic.py#L1288

In your use case, replacing it with [`build_analyzer`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) might be more appropriate.
